Basic/Invaders/Assets/Scripts/InvadersGame.cs

```
@@ -268,90 +290,109 @@ private void UpdateGameTimer()
 
-            if (foundEdge)
+            // If the enemies reached the either side of the boundaries, then change the movement direction
+            // And move them to the next row below
+            if ((enemiesResultFlags & UpdateEnemiesResultFlags.ReachedHorizontalBoundary) != 0)
             {
-                m_AlienDirection.Value = -m_AlienDirection.Value;
+                m_EnemyMovingDirection.Value = -m_EnemyMovingDirection.Value;
                 m_TickPeriodic.Value *= 0.95f; // get faster
 
-                var aliensCount = m_Aliens.Count;
-                for (var index = 0; index < aliensCount; index++)
+                var enemiesCount = m_Enemies.Count;
+                for (var index = 0; index < enemiesCount; index++)
                 {
-                    var alien = m_Aliens[index];
-                    alien.transform.Translate(0, k_AlienVerticalMovementOffset, 0);
+                    var enemy = m_Enemies[index];
+                    enemy.transform.Translate(0, k_EnemyVerticalMovementOffset, 0);
                 }
             }
 
             if (m_Saucer == null)
                 if (Random.Range(0, 1.0f) > m_RandomThresholdForSaucerCreation)
-                    CreateSaucer();
+                    CreateSuperEnemy();
         }
     }
 
-    private bool UpdateShootingEnemies(ref bool foundAlien, ref bool foundEdge)
+    private void UpdateShootingEnemies(ref UpdateEnemiesResultFlags flags)
     {
-        var aliensCount = m_Aliens.Count;
-        for (var index = 0; index < aliensCount; index++)
+        flags = UpdateEnemiesResultFlags.None;
+        var enemiesCount = m_Enemies.Count;
+        for (var index = 0; index < enemiesCount; index++)
         {
-            var alien = m_Aliens[index];
-            Assert.IsTrue(alien);
-
-            if (alien.score > 100)
+            var enemy = m_Enemies[index];
+            Assert.IsTrue(enemy != null);
+            if (m_Enemies == null)
+            {
+                continue;
```

This should never happen as we assert above. And it is a silent fail? Maybe throw an exception here?